yjit_codegen.c

```diff
@@ -1430,6 +1435,130 @@ gen_set_ivar(jitstate_t *jit, ctx_t *ctx, const int max_chain_depth, VALUE compt
     VALUE comptime_val_klass = CLASS_OF(comptime_receiver);
     const ctx_t starting_context = *ctx; // make a copy for use with jit_chain_guard
 
+    // If the class uses the default allocator, instances should all be T_OBJECT
+    // NOTE: This assumes nobody changes the allocator of the class after allocation.
+    //       Eventually, we can encode whether an object is T_OBJECT or not
+    //       inside object shapes.
+    if (!RB_TYPE_P(comptime_receiver, T_OBJECT) ||
+            rb_get_alloc_func(comptime_val_klass) != rb_class_allocate_instance) {
+        // General case. Call rb_ivar_get(). No need to reconstruct interpreter
+        // state since the routine never raises exceptions or allocate objects
+        // visibile to Ruby.
+        // VALUE rb_ivar_set(VALUE obj, ID id, VALUE val)
+        ADD_COMMENT(cb, "call rb_ivar_set()");
+        mov(cb, C_ARG_REGS[0], REG0);
+        mov(cb, C_ARG_REGS[1], imm_opnd((int64_t)ivar_name));
+        mov(cb, C_ARG_REGS[2], ctx_stack_pop(ctx, 1));
+        call_ptr(cb, REG1, (void *)rb_ivar_set);
+
+        if (!reg0_opnd.is_self) {
+            (void)ctx_stack_pop(ctx, 1);
+        }
+        // FIXME: setting an ivar pushes the same value back on the stack, so we shouldn't
+        // pop and push.
+        // Push the ivar on the stack
+        x86opnd_t out_opnd = ctx_stack_push(ctx, TYPE_UNKNOWN);
+        mov(cb, out_opnd, RAX);
+
+        // Jump to next instruction. This allows guard chains to share the same successor.
+        jit_jump_to_next_insn(jit, ctx);
+        return YJIT_END_BLOCK;
+    }
+
+    // ID for the name of the ivar
+    ID id = ivar_name;
+    struct rb_iv_index_tbl_entry *ent;
+    struct st_table *iv_index_tbl = ROBJECT_IV_INDEX_TBL(comptime_receiver);
+
+    // Lookup index for the ivar the instruction loads
+    if (iv_index_tbl && rb_iv_index_tbl_lookup(iv_index_tbl, id, &ent)) {
+        uint32_t ivar_index = ent->index;
+
```

Comment:
Do we also need a frozen check?

Reply:
🤦 ya we probably do.  I'm going to write a test for it though

Reply:
Fixed and added tests.